clients/lsp-csharp.el: implement commands to run tests

razzmatazz opened the pull request:
* Add `lsp-csharp-run-test-in-buffer`
 
* Add `lsp-csharp-run-test-at-point`

* Add `lsp-csharp-run-all-tests-in-buffer`

---

This is another PR that is waiting for a new `omnisharp-roslyn` server release but is pending review as well.

I am not 100% if that is the preferred interface to run tests in `lsp-mode`. Does anyone know how other languages implement this?

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+          existing-buffer)
+      (let ((buffer (get-buffer-create lsp-csharp-test-run-buffer-name)))
+        (with-current-buffer buffer
+          (setq default-directory project-root-dir)
```
This has no effect, because `compilation-mode` is a minor-mode and as such deletes all local bindings within the buffer upon being enabled.
```emacs-lisp
(with-temp-buffer (setq-local default-directory t) (compilation-mode) default-directory) ; = whatever the default value of `default-directory` is. `setq-local` is for illustration here only, since that variable is defined buffer-local
```

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+    (if existing-buffer
+        (progn
+          (with-current-buffer existing-buffer
+            (setq buffer-read-only nil)
```
usually, this is performed in `let` bound `buffer-read-only`

nbfalcon replied:
This would be `((inhibit-read-only t))`, no?

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+
+(defun lsp-csharp--test-message (message)
+  "Emit a MESSAGE to lsp-csharp test run buffer."
+  (let ((existing-buffer (get-buffer lsp-csharp-test-run-buffer-name)))
```
This should be `when-let`, allowing you to drop the `if`.

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+  (let ((existing-buffer (get-buffer lsp-csharp-test-run-buffer-name)))
+    (if existing-buffer
+        (with-current-buffer existing-buffer
+          (setq buffer-read-only nil)
```
You can use `(let ((inhibit-read-only t)) ...)` which would be much cleaner.

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+          (goto-char (point-max))
+          (insert message)
+          (insert "\n")
+          (setq buffer-read-only t))))))
```
With the above, you can drop this.

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+        (seq-doseq (stdout-line standard-output)
+          (lsp-csharp--test-message stdout-line)))
+
+      (unless (= (seq-length standard-error) 0)
```
nit: there is `seq-empty-p`

nbfalcon replied:
`seq-empty-p`

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+        (with-current-buffer existing-buffer
+          (setq buffer-read-only nil)
+          (goto-char (point-max))
+          (insert message)
```
Why not `(insert message "\n")`?

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+
+(defun lsp-csharp--test-message (message)
+  "Emit a MESSAGE to lsp-csharp test run buffer."
+  (let ((existing-buffer (get-buffer lsp-csharp-test-run-buffer-name)))
+    (if existing-buffer
```
nit: use when-let

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+          (compilation-mode)
+          buffer))))
+
+  (if present-buffer
```
Use `when` or `and` if there is no else branch.

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+  "Emit a MESSAGE to lsp-csharp test run buffer."
+  (let ((existing-buffer (get-buffer lsp-csharp-test-run-buffer-name)))
+    (if existing-buffer
+        (with-current-buffer existing-buffer
```
nit: typically, this is performed in save-excursion block. There is a similar code in `dap--print-to-output-buffer` to preserve the point depending on point position.

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+    (unless passed
+      (lsp-csharp--test-message error-message)
+
+      (if error-stack-trace
```
`when`

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+  (interactive)
+  (lsp-csharp--reset-test-buffer t)
+  (lsp-send-execute-command "omnisharp/runTestMethod"
+                            (vector (ht ("textDocumentUri" (lsp--buffer-uri))
```
You can use a `plist` here: `(vector :textDocumentUri (lsp--buffer-uri))`...

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+
+PRESENT-BUFFER will make the buffer be presented to the user."
+  (let ((existing-buffer (get-buffer lsp-csharp-test-run-buffer-name))
+        (project-root-dir (lsp--suggest-project-root)))
```
you should use lsp-workspace-root

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+  (interactive)
+  (lsp-csharp--reset-test-buffer t)
+  (lsp-send-execute-command "omnisharp/runTestMethod"
+                            (vector (ht ("textDocumentUri" (lsp--buffer-uri))))))
```
Same as the above.

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+      (if error-stack-trace
+          (lsp-csharp--test-message error-stack-trace))
+
+      (unless (= (seq-length standard-output) 0)
```
`seq-empty-p`

yyoncho replied:
nit: there is seq-empty-p

nbfalcon reviewed (approved):
Mainly nits from me, but otherwise LGTM. Someone who uses the csharp-ls will have to test this, though.

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+          existing-buffer)
+      (let ((buffer (get-buffer-create lsp-csharp-test-run-buffer-name)))
+        (with-current-buffer buffer
+          (setq default-directory project-root-dir)
```
why this is needed?

razzmatazz replied:
When in compliation mode file paths in error messages and such can be clicked-on/activated via keyboard -- as the paths are not normalized at FS root (`/`) -- this helps it to resolve the actual file. At least this is the behaviour in doom-emacs configuration I have -- not sure if that is the standard behaviour.

I.e. test run buffer contains output from the compiler as well (warnings+errors), not just the actual output from tests.

razzmatazz replied:
There might be a better way, right.

nbfalcon replied:
```emacs-lisp
(with-current-buffer (get-buffer-create lsp-csharp-test-buffer-name)
  (compilation-mode)
  (read-only-mode 1)
  (setq-local default-directory (lsp-workspace-root))
  (let ((inhibit-read-only t))
    (erase-buffer)
    (insert message "\n"))
  (current-buffer))
```

nbfalcon replied:
Another thing that could be done is to make a special major-mode for this, deriving from compilation-mode so that users can set-up custom keybindings. I.e.
```emacs-lisp
(define-derived-mode lsp-csharp-test-mode compilation-mode
  "Major mode for running `lsp-csharp` tests"
  (read-only-mode 1)
  (setq-local default-directory (lsp-workspace-root))
```
And then the first three lines (compilation-mode ...) could simply be `(lsp-csharp-test-mode)`.

yyoncho reviewed (commented):
Looks good. Some nits. I will need one more pass.

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+        (with-current-buffer buffer
+          (setq default-directory project-root-dir)
+          (compilation-mode)
+          buffer))))
```
You can add `(read-only-mode 1)` here.

nbfalcon commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,99 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+        (project-root-dir (lsp--suggest-project-root)))
+    (if existing-buffer
+        (progn
+          (with-current-buffer existing-buffer
```
You can unify the set-up, which would make it more idiomatic:
```emacs-lisp
(with-current-buffer (get-buffer-create ...)
  (compilation-mode)
  (read-only-mode 1)
  (setq-local default-directory ...)
  (let ((inhibit-read-only t))
    (insert ...) ...))
 ```

razzmatazz commented:
> Looks good. Some nits. I will need one more pass.

Thanks @yyoncho and @nbfalcon !

I still suspect the code might change as I am not sure if the PR on server side should be merged or if I can reuse existing `o#/*` endpoints with extra lisp code. Thus it will need another pass some time later, especially as server version is released every month or so..

razzmatazz commented:
@nbfalcon @yyoncho I have updated the code according to your comments.

One thing that has changed is that I have:
- added a new interactive command lsp-csharp-run-test-method-in-buffer (that does completing-read to select a test to run)

- rewrote the code to not use the `omnisharp/runTestMethod` server method (which I would prefer not to merge) -- rather to use the legacy protocol of omnisharp-roslyn that retrieves the list of code elements for the buffer to figure out which tests to run.

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,188 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+             (propertize
+              (upcase outcome)
+              'font-lock-face (if passed
+                                  '(:foreground "green" :weight bold)
```
consider using error and success faces.

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,188 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+  "Flatten the omnisharp:CodeElement tree in ELEMENTS matching PREDICATE."
+  (let ((results nil))
+    (lsp-csharp--inspect-code-elements-recursively (lambda (el)
+                                                     (if (funcall predicate el)
```
nit: use when instead of if

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,188 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+
+(defun lsp-csharp--code-element-stack-at-point ()
+  "Return omnisharp:CodeElement stack at point as a list."
+  (let ((pos-line (- (line-number-at-pos) 1))
```
can you get these from `lsp--cur-position` result? It handles virtual buffers and in the future will handle multi-byte chars in the proper way.

yyoncho commented on clients/lsp-csharp.el:
```diff
@@ -235,6 +241,188 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
+(lsp-defun lsp-csharp--code-element-test-method-p (element)
+  "Return test method name and test framework for a given ELEMENT."
+  (-when-let* (((&omnisharp:CodeElement :kind :properties) element)
+               ((&hash "testMethodName" test-method-name
```
You should not use &hash for stuff that comes from the server, since setting lsp-use-plists will turn server results into plists. Use lsp-get or define new interface.

yyoncho reviewed (commented):
Looks good! Mostly trivial nitpicking.

razzmatazz commented:
@yyoncho I think I took care of all of your nitpicks. 

One thing I wasn't sure how to approach is this line:
- https://github.com/emacs-lsp/lsp-mode/pull/2574/files#diff-d1c03bcd1ccd0772a57b998e7b7278be607a4edefc38c27209c371d8a2504bffR321

I had to convert a `list` (cons) to a vector with `vconcat` so json serialization wouldn't balk at me

yyoncho commented:
> I had to convert a `list` (cons) to a vector with `vconcat` so json serialization wouldn't balk at me.

The alternative of `vconcat` is `(apply #'vector` but I guess `vconcat` is good enough.

yyoncho commented:
As a side note, you might want to update the C# docs to include the commands(+ gifs eventually). You can create a page like `docs/manual-language-docs/lsp-sqls.md` and it will be attached to automatically created docs.

razzmatazz commented:
> As a side note, you might want to update the C# docs to include the commands(+ gifs eventually). You can create a page like `docs/manual-language-docs/lsp-sqls.md` and it will be attached to automatically created docs.

Great idea! Will do, for the other PRs as well.

yyoncho commented:
Good, the code looks fine to me, so let me know when you are good to be merged.

razzmatazz commented:
This good to merge now -- the related LSP server PR is going to merged soon and in some time the new version will be released:
- https://github.com/OmniSharp/omnisharp-roslyn/pull/2099

razzmatazz commented:
@yyoncho the corresponding [server-side PR](https://github.com/OmniSharp/omnisharp-roslyn/pull/2099) has been merged, thus this can too :)

yyoncho commented:
Thank you!
